File: display/theme.lua
-- This module holds the look. Every other module draws through it, so the
-- palette, the type scale, and the drawing shapes are defined in one place.
local theme = {}

-- The display draws in one space 1080 rows tall. libass scales the whole
-- overlay to the real output, so the same layout serves 720, 1080, and 4K
-- with no branch.
-- The width follows the real surface's own ratio, so a canvas pixel is
-- square. A fixed 1920 canvas on a 21:9 screen stretched every vector
-- drawing by a third and pulled every margin inside where it belonged. A
-- 16:9 surface gives 1920, the width this space always held, so a 16:9
-- screen draws every number it drew before.
theme.canvas = { w = 1920, h = 1080 }

-- update_canvas takes the width from the surface mpv reports. Every module
-- reads theme.canvas at draw time, because a width captured at load would
-- hold the width of another screen. main calls this before it tells the
-- modules the size changed.
function theme.update_canvas()
  local d = mp.get_property_native("osd-dimensions")
  if not d or not d.w or d.w <= 0 or not d.h or d.h <= 0 then
    return
  end
  theme.canvas.w = math.floor(theme.canvas.h * d.w / d.h + 0.5)
end

-- osd_scale reads how the canvas maps to the real screen. sx maps a canvas x to
-- real pixels, and sy maps a canvas y. It is nil until a frame has rendered and
-- mpv reports a size.
-- The canvas takes its width from the same surface, so the two factors come
-- out equal. A caller that places a bitmap still reads the axis it means.
function theme.osd_scale()
  local d = mp.get_property_native("osd-dimensions")
  if not d or not d.w or d.w <= 0 or not d.h or d.h <= 0 then
    return nil
  end
  return { sx = d.w / theme.canvas.w, sy = d.h / theme.canvas.h, w = d.w, h = d.h }
end

-- An ASS color is BGR hex, the byte order libass reads, not RGB. The values
-- are liken brand tokens from the brand theme's liken.css. The accent fill is
-- the dark-scheme lichen green --link, the text is --ink, and the muted grey
-- is --ink-muted. The bar track is the light-scheme --link, a deep green dark
-- enough that the bright elapsed fill reads over it.
theme.color = {
  text = "&HE8E8E8&",
  fill = "&H9AC4B4&",
  track = "&H3A5D4A&",
  muted = "&HADA6A0&",
  playhead = "&H9AC4B4&",
  shadow = "&H000000&",
}

-- The global fade factor, from 0 clear to 1 full. drawing and theme.text scale
-- every alpha by it, so the whole OSD fades as one. At 1 the alpha is unchanged.
theme.fade = 1

function theme.set_fade(v)
  theme.fade = v
end

-- faded scales one ASS alpha byte by the fade factor. The byte runs 00 opaque to
-- FF transparent, and out = 255 - round(fade * (255 - in)). At fade 1 the output
-- byte equals the input, so a full OSD draws as it does with no fade at all.
local function faded(alpha)
  local hex = alpha:match("&H(%x+)&")
  if not hex then
    return alpha
  end
  local t = tonumber(hex, 16)
  local out = 255 - math.floor(theme.fade * (255 - t) + 0.5)
  return string.format("&H%02X&", out)
end

-- faded_alpha is faded for a module that sets an alpha inline. It fades an inline
-- override with the rest of the OSD, so a segment with its own alpha does not hold
-- that alpha while everything around it fades.
function theme.faded_alpha(alpha)
  return faded(alpha)
end

-- The fade timing lives here because two things fade on clocks of their own,
-- the OSD and the volume indicator, and the two must look the same. A fade
-- takes fade_in_ms to reach full and fade_out_ms to reach clear, and the out
-- is longer than the in, so anything on the display leaves more slowly than
-- it arrives.
theme.fade_in_ms = 350
theme.fade_out_ms = 600
-- A fade steps on this period, about sixty times a second, and requests a
-- redraw on each step.
theme.fade_tick = 1 / 60
-- An element the display summons for one action leaves this many seconds
-- after the last one. The OSD and the volume indicator wait out the same
-- window, each on its own timer.
theme.idle_hide = 4

-- An ASS alpha runs from 00, opaque, to FF, transparent.
theme.alpha = {
  opaque = "&H00&",
  subdued = "&H80&",
  track = "&H50&",
  dim = "&HA8&",
  panel = "&H14&",
  highlight = "&H30&",
}

-- The type scale, in canvas pixels. The sizes are large enough to read from
-- a couch at 1080.
theme.type = {
  title = 64,
  label = 40,
  small = 34,
  tiny = 28,
}

-- The layout margins and the shared rows. x is the side margin every
-- flush-left and flush-right element keeps. y is the top margin, and by
-- symmetry the bottom one: the clock, the header, and the activity line
-- hang from it, and the identity block and the preview legend stand the
-- same distance off the bottom edge. bar_y is the scrubber bar's center
-- line, where the image counter also sits. panel_bottom is the baseline a
-- chooser or an adjuster panel grows upward from. They live here because
-- two modules that share a row must read one number, not keep two copies
-- that happen to agree.
theme.margin = {
  x = 140,
  y = 90,
}
theme.bar_y = 904
theme.panel_bottom = 876
-- line_pitch is the drop from one line of the top-right column to the next.
-- The clock hangs at the top margin, the activity line one pitch under it,
-- and the volume row one pitch under that, so the three read as a column and
-- no two of them touch.
theme.line_pitch = theme.type.small + 12

-- The whole display draws in one family. libass resolves it through fontconfig,
-- and the player image installs the font, so the text renders the same on every
-- machine. With no installed match, libass falls back and the look drifts.
theme.font = "Source Sans 3"

-- Place a shape at its top-left with an7 and pos, so the path points are the
-- shape's own local box. p1 is the ASS vector drawing mode.
-- The border is off by default, so a shape draws its fill alone. A shape that
-- passes a width and a color draws an outline whose alpha matches the fill
-- alpha, so the border and the fill dim together.
local function drawing(x, y, color, alpha, path, bord, bordcolor, blur)
  bord = bord or 0
  bordcolor = bordcolor or color
  blur = blur or 0
  local a = faded(alpha)
  return string.format(
    "{\\an7\\pos(%.2f,%.2f)\\bord%.2f\\3c%s\\3a%s\\shad0\\blur%.2f\\1c%s\\1a%s\\p1}%s{\\p0}",
    x, y, bord, bordcolor, a, blur, color, a, path
  )
end

function theme.rect(x, y, w, h, color, alpha)
  local path = string.format("m 0 0 l %.2f 0 l %.2f %.2f l 0 %.2f", w, w, h, h)
  return drawing(x, y, color, alpha, path)
end

-- The heights of the scrim's top band and bottom band, in canvas pixels.
theme.scrim_top_h = 410
theme.scrim_bottom_h = 480
-- The scrim's dark plateau, as an ASS alpha. 0 is opaque, 255 is clear.
theme.scrim_edge_alpha = 0x34
-- The dark plateau covers this fraction of the scrim height, at the screen
-- edge, over the text. The blur softens its inner edge.
theme.scrim_solid = 0.66
-- How far the blur carries the fade inward, as a fraction of the scrim height.
theme.scrim_reach = 0.3

-- theme.scrim draws the dark gradient behind a cluster of text, so the text
-- reads against one background whatever the frame behind it. edge names the
-- dark side, top or bottom, and the far side fades to clear.
--
-- The scrim is one blurred shape, not a stack of translucent bands. Tiled bands
-- meet at shared edges, and a wide film that squishes the tall canvas lands
-- those edges sub-pixel, where they alias into fine lines. One blurred shape has
-- no interior edges, so it stays smooth at any output scale. A dark plateau
-- covers the text at the screen edge, and the blur carries its inner edge to
-- clear. The rectangle bleeds past the screen on every side but the inner one
-- by the blur width, so the blur there falls off the screen. Only the inner
-- edge fades, and the very edge stays fully dark.
function theme.scrim(x, y, w, h, edge)
  local solid = h * theme.scrim_solid
  local blur = h * theme.scrim_reach
  local ry
  if edge == "bottom" then
    ry = y + h - solid
  else
    ry = y - blur
  end
  local rh = solid + blur
  local rx = x - blur
  local rw = w + 2 * blur
  local path = string.format("m 0 0 l %.2f 0 l %.2f %.2f l 0 %.2f", rw, rw, rh, rh)
  return drawing(
    rx, ry, theme.color.shadow, string.format("&H%02X&", theme.scrim_edge_alpha), path, 0, nil, blur
  )
end

local function rounded_path(w, h, r)
  r = math.min(r, w / 2, h / 2)
  return string.format(
    "m %.2f 0 l %.2f 0 b %.2f 0 %.2f 0 %.2f %.2f l %.2f %.2f "
      .. "b %.2f %.2f %.2f %.2f %.2f %.2f l %.2f %.2f "
      .. "b 0 %.2f 0 %.2f 0 %.2f l 0 %.2f b 0 0 0 0 %.2f 0",
    r, w - r, w, w, w, r, w, h - r,
    w, h, w, h, w - r, h, r, h,
    h, h, h - r, r, r
  )
end

function theme.rounded_rect(x, y, w, h, r, color, alpha)
  return drawing(x, y, color, alpha, rounded_path(w, h, r))
end

-- A menu panel: a faint dark fill inside a solid green border, so a chooser or
-- an adjuster reads as one surface over the video. The border is liken's green,
-- and the fill dims the video behind the text without hiding it.
function theme.panel(x, y, w, h)
  return string.format(
    "{\\an7\\pos(%.2f,%.2f)\\bord2\\3c%s\\3a&H00&\\shad0\\1c%s\\1a%s\\p1}%s{\\p0}",
    x, y, theme.color.fill, theme.color.shadow, theme.alpha.panel, rounded_path(w, h, 14)
  )
end

-- A pointy-top regular hexagon in a 2r box, center at (r, r). r is the
-- distance from the center to a vertex. The 0.866 is cos(30 degrees), the
-- half-width of a pointy-top hexagon.
-- The shape needs no counter-stretch. The canvas holds the screen's own
-- ratio, so a hexagon drawn regular on the canvas lands regular on the
-- screen.
local function hexagon_path(r)
  local a = 0.8660254 * r
  return string.format(
    "m %.2f 0 l %.2f %.2f l %.2f %.2f l %.2f %.2f l %.2f %.2f l %.2f %.2f",
    r, r + a, 0.5 * r, r + a, 1.5 * r, r, 2 * r, r - a, 1.5 * r, r - a, 0.5 * r
  )
end

function theme.hexagon(x, y, r, color, alpha, bord, bordcolor)
  return drawing(x, y, color, alpha, hexagon_path(r), bord, bordcolor)
end

-- theme.shape draws one ASS path the caller supplies, for a mark the named
-- shapes above do not cover, such as the volume glyph. The path is in the
-- shape's own local box, and the shape takes the same fade every other shape
-- takes. A border width and a color draw an outline, so a mark over another
-- mark in the same color reads apart from it.
function theme.shape(x, y, path, color, alpha, bord, bordcolor)
  return drawing(x, y, color, alpha, path, bord, bordcolor)
end

-- The scrim holds the contrast, so the label draws flat, with no outline.
function theme.text(x, y, s, size, color, an, alpha)
  alpha = alpha or theme.alpha.opaque
  return string.format(
    "{\\an%d\\pos(%.2f,%.2f)\\fn%s\\fs%d\\bord0\\shad0\\1c%s\\1a%s}%s",
    an, x, y, theme.font, size, color, faded(alpha), s
  )
end

return theme

File: display/src/theme.rs
//! This module holds the look. Every other module draws through it, so the
//! palette, the type scale, and the geometry constants are defined in one
//! place.

use std::time::Duration;

use iced::Color;

/// The display draws in one space 1080 rows tall. The canvas module scales
/// the whole layer to the real output, so the same layout serves 720, 1080,
/// and 4K with no branch.
pub const CANVAS_HEIGHT: f32 = 1080.0;

/// The width a 16:9 surface gives, which is the width this space always
/// held. The canvas module takes the width from the surface's own ratio, so
/// a canvas pixel is square.
pub const CANVAS_WIDTH: f32 = 1920.0;

/// The values are liken brand tokens from the brand theme's `liken.css`. The
/// accent fill is the dark-scheme lichen green `--link`, the text is
/// `--ink`, and the muted grey is `--ink-muted`. The bar track is the
/// light-scheme `--link`, a deep green dark enough that the bright elapsed
/// fill reads over it.
pub mod color {
    use super::Color;

    /// The colour of body text.
    pub fn text() -> Color {
        liken_iced::palette::dark().ink
    }

    /// The accent, and the colour of the elapsed fill.
    pub fn fill() -> Color {
        liken_iced::palette::dark().link
    }

    /// The bar's unplayed track.
    pub fn track() -> Color {
        liken_iced::palette::light().link
    }

    /// The colour of text that reads under body text.
    pub fn muted() -> Color {
        liken_iced::palette::dark().ink_muted
    }

    /// The playhead, which takes the accent.
    pub fn playhead() -> Color {
        fill()
    }

    /// The scrim and every shadow.
    pub const SHADOW: Color = Color::BLACK;
}

/// How much of the ground each surface covers, from 1 opaque to 0 clear.
/// Every value states the ASS alpha byte the look was drawn with, so the
/// display covers what the Lua display covered, and nothing downstream reads
/// a byte whose 0 means opaque.
pub mod alpha {
    use super::opacity;

    pub const OPAQUE: f32 = opacity(0x00);
    pub const SUBDUED: f32 = opacity(0x80);
    pub const TRACK: f32 = opacity(0x50);
    pub const DIM: f32 = opacity(0xA8);
    pub const PANEL: f32 = opacity(0x14);
    pub const HIGHLIGHT: f32 = opacity(0x30);
    /// The scrim's dark plateau, and the volume row's own surface.
    pub const SCRIM_EDGE: f32 = opacity(0x34);
}

/// One ASS alpha byte as the fraction of the ground it covers. An ASS alpha
/// runs from 00, opaque, to FF, transparent.
pub const fn opacity(alpha: u8) -> f32 {
    1.0 - alpha as f32 / 255.0
}

/// One colour at one coverage. Every drawing call takes a colour that already
/// carries the fraction it covers.
pub fn at(color: Color, alpha: f32) -> Color {
    Color { a: alpha, ..color }
}

/// One colour under the fade. The fade scales every alpha the display draws,
/// so the whole layer fades as one. At a fade of 1 the colour covers what it
/// states.
pub fn faded(color: Color, fade: f32) -> Color {
    Color {
        a: color.a * fade,
        ..color
    }
}

/// The fade timing lives here because two things fade on clocks of their
/// own, the OSD and the volume indicator, and the two must look the same. A
/// fade takes `FADE_IN` to reach full and `FADE_OUT` to reach clear, and the
/// out is longer than the in, so anything on the display leaves more slowly
/// than it arrives.
pub const FADE_IN: Duration = Duration::from_millis(350);
pub const FADE_OUT: Duration = Duration::from_millis(600);

/// A fade steps on this period, about sixty times a second, and redraws on
/// each step.
pub const FADE_TICK: Duration = Duration::from_nanos(16_666_667);

/// An element the display summons for one action leaves this many seconds
/// after the last one. The OSD and the volume indicator wait out the same
/// window, each on its own timer.
pub const IDLE_HIDE: Duration = Duration::from_secs(4);

/// The type scale, in canvas pixels. The sizes are large enough to read from
/// a couch at 1080. Each number is a line box, the measure the display
/// states as an ASS `\\fs`, and [`type_size`] turns one into the size the
/// toolkit takes.
pub mod type_scale {
    pub const TITLE: f32 = 64.0;
    pub const LABEL: f32 = 40.0;
    pub const SMALL: f32 = 34.0;
    pub const TINY: f32 = 28.0;
}

/// The face's metric: its bounding height over its em, 1326 units over 1000.
/// The bounding height is `usWinAscent` plus `usWinDescent` and the em is
/// `unitsPerEm`, from the `OS/2` and `head` tables of
/// `SourceSans3-Regular.otf`, the face the brand crate carries and both
/// renderers draw.
///
/// libass scales a face so that its bounding height fills the size an ASS
/// `\\fs` states, so one `\\fs` number states two measures: the line box the
/// text draws in, in canvas pixels, and the type size, which is that box
/// divided by this metric. A layout measure the display writes in `\\fs`
/// units is a canvas measure here and passes through unchanged; only a type
/// size goes through the metric.
const FACE_METRIC: f32 = 1326.0 / 1000.0;

/// The type size that draws in a line box `height` canvas pixels tall. A
/// line's anchor falls on that box in both renderers, so a line placed by its
/// top or its bottom puts its baseline where libass puts it.
pub fn type_size(height: f32) -> f32 {
    height / FACE_METRIC
}

/// The side margin every flush-left and flush-right element keeps.
pub const MARGIN_X: f32 = 96.0;

/// The top margin, which the header, the clock, and the volume row all
/// measure down from.
pub const MARGIN_Y: f32 = 90.0;

/// The scrubber bar's center line, which the image counter shares.
pub const BAR_Y: f32 = 904.0;

/// The baseline a chooser or an adjuster panel grows upward from.
pub const PANEL_BOTTOM: f32 = 876.0;

/// The pitch of one line in the top-right column.
pub const LINE_PITCH: f32 = type_scale::SMALL + 12.0;

/// The heights of the scrim's top band and bottom band, in canvas pixels.
pub const SCRIM_TOP_HEIGHT: f32 = 410.0;
pub const SCRIM_BOTTOM_HEIGHT: f32 = 480.0;

/// The dark plateau covers this fraction of the scrim height, at the screen
/// edge, over the text. The fade softens its inner edge.
pub const SCRIM_SOLID: f32 = 0.66;

/// How far the fade carries the plateau inward, as a fraction of the scrim
/// height.
pub const SCRIM_REACH: f32 = 0.3;

#[cfg(test)]
mod tests {
    use super::*;

    /// The four tokens the palette reads out of `liken.css`.
    #[test]
    fn the_palette_holds_the_colours() {
        let hex = |color: Color| {
            let byte = |channel: f32| (channel * 255.0).round() as u8;
            format!(
                "#{:02X}{:02X}{:02X}",
                byte(color.r),
                byte(color.g),
                byte(color.b)
            )
        };

        assert_eq!(hex(color::text()), "#E8E8E8");
        assert_eq!(hex(color::fill()), "#B4C49A");
        assert_eq!(hex(color::track()), "#4A5D3A");
        assert_eq!(hex(color::muted()), "#A0A6AD");
        assert_eq!(hex(color::playhead()), "#B4C49A");
        assert_eq!(hex(color::SHADOW), "#000000");
    }

    /// Every coverage is the ASS alpha byte the look was drawn with, read as
    /// the fraction of the ground it covers.
    #[test]
    fn every_coverage_is_the_ass_alpha_byte_it_was_drawn_with() {
        for (coverage, byte) in [
            (alpha::OPAQUE, 0x00),
            (alpha::SUBDUED, 0x80),
            (alpha::TRACK, 0x50),
            (alpha::DIM, 0xA8),
            (alpha::PANEL, 0x14),
            (alpha::HIGHLIGHT, 0x30),
            (alpha::SCRIM_EDGE, 0x34),
        ] {
            assert_eq!(coverage, 1.0 - f32::from(byte as u8) / 255.0, "{byte:#04x}");
        }
        assert_eq!(alpha::OPAQUE, 1.0);
        assert_eq!(opacity(0xFF), 0.0);
        assert!((alpha::SUBDUED - 0.498).abs() < 0.001);
        assert!((alpha::TRACK - 0.686).abs() < 0.001);
        assert!((alpha::DIM - 0.341).abs() < 0.001);
        assert!((alpha::PANEL - 0.922).abs() < 0.001);
        assert!((alpha::HIGHLIGHT - 0.812).abs() < 0.001);
        assert!((alpha::SCRIM_EDGE - 0.796).abs() < 0.001);
    }

    /// The display states a size as an ASS `\\fs`, which is the line box. The
    /// toolkit takes the type size, which is that box through the metric.
    #[test]
    fn a_type_size_is_its_line_box_through_the_face_metric() {
        assert!((type_size(type_scale::SMALL) - 25.641).abs() < 0.001);
        assert!((type_size(type_scale::TINY) - 21.116).abs() < 0.001);
        assert!((type_size(type_scale::LABEL) - 30.166).abs() < 0.001);
        assert!((type_size(type_scale::TITLE) - 48.265).abs() < 0.001);
    }

    #[test]
    fn a_faded_colour_keeps_its_channels_and_scales_its_alpha() {
        let full = faded(at(color::text(), alpha::OPAQUE), 1.0);
        assert_eq!(
            (full.r, full.g, full.b),
            (color::text().r, color::text().g, color::text().b)
        );
        assert_eq!(full.a, 1.0);
        assert_eq!(faded(at(color::fill(), alpha::OPAQUE), 0.5).a, 0.5);
        assert_eq!(faded(at(color::fill(), alpha::SUBDUED), 0.0).a, 0.0);
        assert!((faded(at(color::fill(), alpha::SUBDUED), 1.0).a - 0.498).abs() < 0.001);
    }
}
